### PR TITLE
Database Refactor: Remove UsedBy/Config from Projects

### PR DESCRIPTION
**Pull request by masnax**
This adds the last chunk of changes needed for the v2 generator, and is currently in use by `Projects`.

* Removes `UsedBy` and `Config` fields from `db.Project`. 
* Moves `db/projects.go` to use the `version=2` generator.
* Updates generator to support a `references` key, allowing for generation of `Config` and `Devices` methods on each entity.
* Adds generated `GetProjectConfig` and `CreateProjectConfig` methods, generated as part of `GetMany` and `Create` generator comments, specified via `references=Config`.
* Adds a `URL` method to `api.Project`.
* Updates `project.ToAPI()` to accept a transaction argument to fill the config field. 
* Uses `api.Project` or explicitly passing in the config type `map[string]string` where we need to access the config.
* adds `projectUsedBy` helper to `lxd/api_projects.go` and changes `projectIsEmpty` to return immediately if we know a project is used, rather than waiting for the result of `projectUsedBy`.

### CODE REVIEW

**Comment by masnax** [comment]
@markylaing The changes to `lxd/db/generate/db/method_v2.go` are likely going to be the final form of ClusterTx method generation for the db generator. We will likely want to avoid using anything in `method.go` and instead use `v2` going forward.

**Comment by tomponline** [comment]
@masnax unit tests are failing im afraid

**Comment by masnax** [comment]
@tomponline sorry about that, just a stubborn file that I deleted, which got resurrected by a rebase :)

File: lxd/api_project.go [hunk 1]
```diff
@@ -191,6 +191,45 @@ func projectsGet(d *Daemon, r *http.Request) response.Response {
...
 }
 
+// projectUsedBy returns a list of URLs for all instances, images, profiles,
+// storage volumes, networks, and acls that use this project.
+func projectUsedBy(tx *db.ClusterTx, project *db.Project) ([]string, error) {
+	instances, err := tx.GetInstanceURIs(db.InstanceFilter{Project: &project.Name})
+	if err != nil {
+		return nil, err
+	}
+	images, err := tx.GetImageURIs(db.ImageFilter{Project: &project.Name})
```
Comment (tomponline) [inline lxd/api_project.go:200]: can you add a blank line between each stanza please

File: lxd/device/proxy.go [hunk 2]
```diff
@@ -126,7 +127,17 @@ func (d *proxy) validateConfig(instConf instance.ConfigReader) error {
 				// Prevent use of NAT mode on non-default projects with networks feature.
 				// This is because OVN networks don't allow the host to communicate directly with
 				// instance NICs and so DNAT rules on the host won't work.
-				p, err := d.state.Cluster.GetProject(projectName)
+				var p *api.Project
+				d.state.Cluster.Transaction(func(tx *db.ClusterTx) error {
+					var err error
```
Comment (tomponline) [inline lxd/device/proxy.go:132]: dont need this line

File: lxd/images.go [hunk 3]
```diff
@@ -2085,14 +2103,23 @@ func pruneLeftoverImages(d *Daemon) {
...
+	var projects []api.Project
 	err := d.cluster.Transaction(func(tx *db.ClusterTx) error {
 		var err error
-		projects, err = tx.GetProjects(db.ProjectFilter{})
+		dbProjects, err := tx.GetProjects(db.ProjectFilter{})
 		if err != nil {
 			return err
 		}
 
+		for _, project := range dbProjects {
```
Comment (tomponline) [inline lxd/images.go:2114]: van you use make here to preallocate a slice of the needed capacity

File: lxd/instance_snapshot.go [hunk 4]
```diff
@@ -233,17 +233,13 @@ func instanceSnapshotsPost(d *Daemon, r *http.Request) response.Response {
 			return err
 		}
 
+		err = project.AllowSnapshotCreation(tx, proj)
```
Comment (tomponline) [inline lxd/instance_snapshot.go:236]: blank line needed

File: lxd/instances_post.go [hunk 5]
```diff
@@ -12,7 +12,7 @@ import (
 	"strings"
 	"time"
 
-	"github.com/dustinkirkland/golang-petname"
+	petname "github.com/dustinkirkland/golang-petname"
```
Comment (tomponline) [inline lxd/instances_post.go:15]: this looks like an unrelated change

File: lxd/network/driver_ovn.go [hunk 6]
```diff
@@ -377,7 +377,17 @@ func (n *ovn) Validate(config map[string]string) error {
 	}
 
 	// Load the project to get uplink network restrictions.
-	p, err := n.state.Cluster.GetProject(n.project)
+	var p *api.Project
+	n.state.Cluster.Transaction(func(tx *db.ClusterTx) error {
+		var err error
```
Comment (tomponline) [inline lxd/network/driver_ovn.go:382]: no need for this line

File: lxd/network/driver_ovn.go [hunk 7]
```diff
@@ -4028,7 +4049,17 @@ func (n *ovn) ForwardCreate(forward api.NetworkForwardsPost, clientType request.
 		}
 
 		// Load the project to get uplink network restrictions.
-		p, err := n.state.Cluster.GetProject(n.project)
+		var p *api.Project
+		n.state.Cluster.Transaction(func(tx *db.ClusterTx) error {
+			var err error
```
Comment (tomponline) [inline lxd/network/driver_ovn.go:4054]: no need for this line

File: lxd/network/driver_ovn.go [hunk 8]
```diff
@@ -4028,7 +4049,17 @@ func (n *ovn) ForwardCreate(forward api.NetworkForwardsPost, clientType request.
 		}
 
 		// Load the project to get uplink network restrictions.
-		p, err := n.state.Cluster.GetProject(n.project)
+		var p *api.Project
+		n.state.Cluster.Transaction(func(tx *db.ClusterTx) error {
```
Comment (tomponline) [inline lxd/network/driver_ovn.go:4053]: no err caught here

File: lxd/project/project.go [hunk 9]
```diff
@@ -103,15 +113,25 @@ func StorageVolumeProjectFromRecord(p *db.Project, volumeType int) string {
 // otherwise the default project name is returned. The second return value is the project's config if non-default
 // project is being returned, nil if not.
 func NetworkProject(c *db.Cluster, projectName string) (string, map[string]string, error) {
-	p, err := c.GetProject(projectName)
+	var project *api.Project
+	err := c.Transaction(func(tx *db.ClusterTx) error {
+		var err error
```
Comment (tomponline) [inline lxd/project/project.go:118]: no need for this line

File: lxd/project/project.go [hunk 10]
```diff
@@ -135,15 +155,25 @@ func NetworkProjectFromRecord(p *db.Project) string {
 // otherwise the default project name is returned. The second return value is the project's config if non-default
 // project is being returned, nil if not.
 func ProfileProject(c *db.Cluster, projectName string) (string, map[string]string, error) {
-	p, err := c.GetProject(projectName)
+	var project *api.Project
+	err := c.Transaction(func(tx *db.ClusterTx) error {
+		var err error
```
Comment (tomponline) [inline lxd/project/project.go:160]: no need for this line

File: lxd/storage_volumes_snapshot.go [hunk 11]
```diff
@@ -116,17 +116,13 @@ func storagePoolVolumeSnapshotsTypePost(d *Daemon, r *http.Request) response.Res
 			return err
 		}
 
+		err = project.AllowSnapshotCreation(tx, proj)
 		return err
```
Comment (tomponline) [inline lxd/storage_volumes_snapshot.go:120]: blank line

File: lxd/api_project.go [hunk 12]
```diff
@@ -648,8 +648,10 @@ func projectChange(d *Daemon, project *api.Project, req api.ProjectPut) response
 		return response.BadRequest(fmt.Errorf("You can't change the features of the default project"))
 	}
 
-	if !projectIsEmpty(project.UsedBy) && featuresChanged {
-		return response.BadRequest(fmt.Errorf("Features can only be changed on empty projects"))
+	if featuresChanged && len(project.UsedBy) > 0 {
+		if len(project.UsedBy) > 1 || !strings.Contains(project.UsedBy[0], "/profiles/default") {
```
Comment (tomponline) [inline lxd/api_project.go:652]: whats the chrck for default about here? needs a commemt.

File: lxd/device/proxy.go [hunk 13]
```diff
@@ -126,7 +127,17 @@ func (d *proxy) validateConfig(instConf instance.ConfigReader) error {
 				// Prevent use of NAT mode on non-default projects with networks feature.
 				// This is because OVN networks don't allow the host to communicate directly with
 				// instance NICs and so DNAT rules on the host won't work.
-				p, err := d.state.Cluster.GetProject(projectName)
+				var p *api.Project
+				d.state.Cluster.Transaction(func(tx *db.ClusterTx) error {
```
Comment (tomponline) [inline lxd/device/proxy.go:131]: err not caught

File: lxd/network/driver_ovn.go [hunk 14]
```diff
@@ -377,7 +377,17 @@ func (n *ovn) Validate(config map[string]string) error {
 	}
 
 	// Load the project to get uplink network restrictions.
-	p, err := n.state.Cluster.GetProject(n.project)
+	var p *api.Project
+	n.state.Cluster.Transaction(func(tx *db.ClusterTx) error {
```
Comment (tomponline) [inline lxd/network/driver_ovn.go:381]: err not caught

File: lxd/network/driver_ovn.go [hunk 15]
```diff
@@ -1727,19 +1737,22 @@ func (n *ovn) setup(update bool) error {
...
-	if err != nil {
-		return fmt.Errorf("Failed to load network restrictions from project %q: %w", n.project, err)
-	}
-
-	// Get project ID.
+	var p *api.Project
 	var projectID int64
-	err = n.state.Cluster.Transaction(func(tx *db.ClusterTx) error {
-		projectID, err = tx.GetProjectID(n.project)
+	n.state.Cluster.Transaction(func(tx *db.ClusterTx) error {
```
Comment (tomponline) [inline lxd/network/driver_ovn.go:1742]: err not caught

File: lxd/network/driver_ovn.go [hunk 16]
```diff
@@ -1727,19 +1737,22 @@ func (n *ovn) setup(update bool) error {
...
+		var err error
```
Comment (tomponline) [inline lxd/network/driver_ovn.go:1743]: line not needed

File: lxd/network/zone/load.go [hunk 17]
```diff
@@ -52,7 +53,16 @@ func Create(s *state.State, projectName string, zoneInfo *api.NetworkZonesPost)
...
-	p, err := s.Cluster.GetProject(projectName)
+	var p *api.Project
+	err = s.Cluster.Transaction(func(tx *db.ClusterTx) error {
+		project, err := tx.GetProject(projectName)
+		if err != nil {
+			return err
+		}
+
+		p, err = project.ToAPI(tx)
+		return err
```
Comment (tomponline) [inline lxd/network/zone/load.go:64]: blank line

File: lxd/network/driver_ovn.go [hunk 18]
```diff
@@ -3050,7 +3063,15 @@ func (n *ovn) InstanceDevicePortValidateExternalRoutes(deviceInstance instance.I
...
 	// Load the project to get uplink network restrictions.
-	p, err = n.state.Cluster.GetProject(n.project)
+	err = n.state.Cluster.Transaction(func(tx *db.ClusterTx) error {
+		project, err := tx.GetProject(n.project)
+		if err != nil {
+			return err
+		}
+
+		p, err = project.ToAPI(tx)
+		return err
```
Comment (tomponline) [inline lxd/network/driver_ovn.go:3073]: blank line

File: lxd/network/driver_ovn.go [hunk 19]
```diff
@@ -377,7 +377,17 @@ func (n *ovn) Validate(config map[string]string) error {
...
+	var p *api.Project
+	n.state.Cluster.Transaction(func(tx *db.ClusterTx) error {
+		var err error
+		project, err := tx.GetProject(n.project)
+		if err != nil {
+			return err
+		}
+
+		p, err = project.ToAPI(tx)
+		return err
```
Comment (tomponline) [inline lxd/network/driver_ovn.go:389]: blank line

File: lxd/instances_post.go [hunk 20]
```diff
@@ -70,7 +70,16 @@ func createFromImage(d *Daemon, r *http.Request, projectName string, req *api.In
...
-			p, err := d.cluster.GetProject(projectName)
+			var p *api.Project
+			err := d.cluster.Transaction(func(tx *db.ClusterTx) error {
+				project, err := tx.GetProject(projectName)
+				if err != nil {
+					return err
+				}
+
+				p, err = project.ToAPI(tx)
+				return err
```
Comment (tomponline) [inline lxd/instances_post.go:81]: blank line

File: lxd/images.go [hunk 21]
```diff
@@ -1773,7 +1782,16 @@ func autoUpdateImage(ctx context.Context, d *Daemon, op *operations.Operation, i
...
-		project, err := d.cluster.GetProject(projectName)
+		var project *api.Project
+		err := d.cluster.Transaction(func(tx *db.ClusterTx) error {
+			p, err := tx.GetProject(projectName)
+			if err != nil {
+				return err
+			}
+
+			project, err = p.ToAPI(tx)
+			return err
```
Comment (tomponline) [inline lxd/images.go:1793]: blank line

File: lxd/images.go [hunk 22]
```diff
@@ -265,7 +265,16 @@ func imgPostInstanceInfo(d *Daemon, r *http.Request, req api.ImagesPost, op *ope
...
-		p, err := d.cluster.GetProject(projectName)
+		var p *api.Project
+		err = d.cluster.Transaction(func(tx *db.ClusterTx) error {
+			project, err := tx.GetProject(projectName)
+			if err != nil {
+				return err
+			}
+
+			p, err = project.ToAPI(tx)
+			return err
```
Comment (tomponline) [inline lxd/images.go:276]: blank line

File: lxd/device/disk.go [hunk 23]
```diff
@@ -349,7 +349,16 @@ func (d *disk) validateEnvironmentSourcePath() error {
...
-		p, err := d.state.Cluster.GetProject(projectName)
+		var p *api.Project
+		err = d.state.Cluster.Transaction(func(tx *db.ClusterTx) error {
+			project, err := tx.GetProject(projectName)
+			if err != nil {
+				return err
+			}
+
+			p, err = project.ToAPI(tx)
+			return err
```
Comment (tomponline) [inline lxd/device/disk.go:360]: blank line

File: lxd/backup.go [hunk 24]
```diff
@@ -77,7 +77,16 @@ func backupCreate(s *state.State, args db.InstanceBackup, sourceInst instance.In
...
-		p, err := s.Cluster.GetProject(sourceInst.Project())
+		var p *api.Project
+		err = s.Cluster.Transaction(func(tx *db.ClusterTx) error {
+			project, err := tx.GetProject(sourceInst.Project())
+			if err != nil {
+				return err
+			}
+
+			p, err = project.ToAPI(tx)
+			return err
```
Comment (tomponline) [inline lxd/backup.go:88]: blank line

File: lxd/project/project.go [hunk 25]
```diff
@@ -71,7 +72,16 @@ func StorageVolumeProject(c *db.Cluster, projectName string, volumeType int) (st
...
-	project, err := c.GetProject(projectName)
+	var project *api.Project
+	err := c.Transaction(func(tx *db.ClusterTx) error {
+		dbProject, err := tx.GetProject(projectName)
+		if err != nil {
+			return err
+		}
+
+		project, err = dbProject.ToAPI(tx)
+		return err
```
Comment (tomponline) [inline lxd/project/project.go:83]: blank line

File: lxd/project/project.go [hunk 26]
```diff
@@ -103,15 +113,25 @@ func StorageVolumeProjectFromRecord(p *db.Project, volumeType int) string {
...
+	var project *api.Project
+	err := c.Transaction(func(tx *db.ClusterTx) error {
+		var err error
+		dbProject, err := tx.GetProject(projectName)
+		if err != nil {
+			return err
+		}
+
+		project, err = dbProject.ToAPI(tx)
+		return err
```
Comment (tomponline) [inline lxd/project/project.go:125]: blank line

File: lxd/project/project.go [hunk 27]
```diff
@@ -135,15 +155,25 @@ func NetworkProjectFromRecord(p *db.Project) string {
...
```
Comment (tomponline) [inline lxd/project/project.go:167]: bl

**Review by tomponline — CHANGES_REQUESTED** [review]
looks good. ill make another pass on my pc tomorrow.

**Comment by stgraber** [comment]
@tomponline what's the state on this one?

**Comment by tomponline** [comment]
@stgraber waiting for @masnax to make the changes then I'll look again.

**Comment by masnax** [comment]
@tomponline sorry, forgot to mention the changes have been made. Should be good to go.

**Comment by tomponline** [comment]
> @tomponline sorry, forgot to mention the changes have been made. Should be good to go.

Cheers, looking now.

**Review by tomponline — APPROVED** [review]
LGTM!

**Pull request merged**